azure-iot-device/azure/iot/device/common/transport/mqtt/pipeline_stages.py

```diff
@@ -0,0 +1,645 @@
+    """
+    PipelineStage which handles a Shared Key Authentication Provider.  This is
+    likely a temporary stage if we're going to get rid of the two-stage
+    client creation.
+
+    Operations Handled:
+    * IotHub_SetAuthProvider
+
+    Operations Produced:
+    * IotHub_SetAuthProviderArgs
+    * Iot_SetSasToken
+
+    This stage handles IotHub_SetAuthProvider operations.  It parses the connection
+    string into it's constituant parts and generates a sas token to pass down.  It
+    uses IotHub_SetAuthProviderArgs to pass the connection string args and the ca_cert
+    from the Authentication Provider, and it uses Iot_SetSasToken to pass down the
+    generated sas token.  After passing down the args and the sas token, this stage
+    completes the IotHub_SetAuthProvider operation.
+
+    All other operations are passed down.
+    """
+
+    def _run_op(self, op):
+        # TODO: this log practice can be moved to run_op()
+        logger.info("{}({}) handling".format(self.name, op.name))
+        if op.name == pipeline_ops.names.IotHub_SetAuthProvider:
+            auth_provider = op.auth_provider
+            if hasattr(auth_provider, "gateway_hostname"):
+                gateway_hostname = auth_provider.gateway_hostname
+            else:
+                gateway_hostname = None
+            self.run_ops_serial(
+                pipeline_ops.IotHub_SetAuthProviderArgs(
+                    device_id=auth_provider.device_id,
+                    module_id=auth_provider.module_id,
+                    hostname=auth_provider.hostname,
+                    gateway_hostname=gateway_hostname,
+                    ca_cert=auth_provider.ca_cert,
+                ),
+                pipeline_ops.Iot_SetSasToken(sas_token=auth_provider.get_current_sas_token()),
+                callback=op.callback,
+            )
+        else:
+            self.continue_op(op)
+
+
+# TODO: make sure everything that is private is named with underscore
+
+
+class Iot_EnsureConnection(PipelineStage):
+    """
+    This stage is responsible for ensuring that the transport is connected when
+    it needs to be connected, and it's responsible for queueing operations
+    while we're waiting for the connect operation to complete.
+
+    Operations Handled:
+    * Iot_Connect
+    * Iot_Disconnect
+    * all operations with needs_connection set to True (connects if necessary)
+    * all operations (queues while waiting for connection)
+
+    Operations Produced:
+    * Iot_Connect
+
+    Note: this stage will likely be replaced by a more full-featured stage to handle
+    other "block while we're setting something up" operations, such as subscribing to
+    twin responses.  That is another example where we want to ensure some state and block
+    requests until that state is achieved.
+    """
+
+    def __init__(self):
+        PipelineStage.__init__(self)
+        self.connected = False
+        self.queue = deque()
```

Comment:
>deque() [](start = 21, length = 7)

why use this instead of a Queue? Do we really need to generalize Stack/Queue behavior?

Reply:
Because I have a brand new hammer and this is very clearly a nail, or a screw, or, OK, I have no problem changing it.

---
In reply to: [279677490](https://github.com/Azure/azure-iot-sdk-python-preview/pull/87#discussion_r279677490) [](ancestors = 279677490)